chore:  docs kyverno comparison (#1010)
defenseunicorns/pepr: docs/080_faq/README.md

export NODE_OPTIONS="--disable-warning=DEP0040"
```

or

```bash
npx --node-options="--disable-warning=DEP0040" pepr [command]
```

## How does Pepr compare to Kyverno?

Although Pepr and Kyverno have similarities, Pepr is very different than Kyverno. They have very different mission statements. Pepr focuses on making operators as easy as possible. Kyverno focuses on reporting, not building operators.

Similarities:

* Both have Mutating Webhooks that can dynamically change resources before admission
* Both have Validating Webhooks to configure what can/cannot go through admission
* Both provide a way to react to changes to pre-existing cluster resources (ie, resources that have already gone through admission)

Differences:

> * Pepr is more like a "framework" than a tool. In Pepr you create a Pepr [Module](https://docs.pepr.dev/main/user*guide/pepr*modules/). In the Pepr module you define [capabilities](https://docs.pepr.dev/main/user*guide/capabilities/) that enforce / apply desired cluster state.

Author: ```suggestion
* Pepr is more like a "framework" than a tool. In Pepr you create a Pepr [Module](../030_user-guide/020_pepr-modules.md). In the Pepr module you define [capabilities](../030_user-guide/040_capabilities.md) that enforce / apply desired cluster state.
```

* Pepr is written in TypeScript. Kyverno is written in Go.
* Pepr provides the flexibility of a full-fledged, strongly typed programming language to decide what decisions to make based on events happening in the cluster. With Kyverno, you are limited to the constraints of YAML.
* Pepr can be used to reconcile events in order, similar to Kube-Builder or Operator SDK.
* Pepr can apply a CustomResourceDefinition and control cluster state based on that custom resource.

Both Pepr and Kyverno are great tools. Which one to use for your project depends on your use case.

## How do I add custom labels to Pepr's Kubernetes manifests?

> During the build process, custom labels can be added the Kubernetes manifests that Pepr generates based on the Pepr section of the `package.json`. Currently, adding custom labels to `namespace` is supported.

Author: ```suggestion
During the build process, custom labels can be added the `pepr-system` namespace based on the `pepr` section of the `package.json`. Checkout the [Customizing with package.json](../030_user-guide/[120_customization.md#packagejson-configurations-table].
```

> During the build process, custom labels can be added the `pepr-system` namespace based on the `pepr` section of the `package.json`. Checkout the [Customizing with package.json](../030_user-guide/120_customization.md#packagejson-configurations-table).

Author: ```suggestion
During the build process, custom labels can be added the `pepr-system` namespace based on the `package.json`. Checkout the [Customizing with package.json](../030_user-guide/120_customization.md#packagejson-configurations-table).
```

> During the build process, custom labels can be added the `pepr-system` namespace based on the `pepr` section of the `package.json`. Checkout the [Customizing with package.json](../030_user-guide/[120_customization.md#packagejson-configurations-table].

Author: ```suggestion
During the build process, custom labels can be added the `pepr-system` namespace based on the `pepr` section of the `package.json`. Checkout the [Customizing with package.json](../030_user-guide/[120_customization.md#packagejson-configurations-table).
```

> During the build process, custom labels can be added the `pepr-system` namespace based on the `pepr` section of the `package.json`. Checkout the [Customizing with package.json](../030_user-guide/[120_customization.md#packagejson-configurations-table).

Author: ```suggestion
During the build process, custom labels can be added the `pepr-system` namespace based on the `pepr` section of the `package.json`. Checkout the [Customizing with package.json](../030_user-guide/120_customization.md#packagejson-configurations-table).
```

Similarities:

* Both have Mutating Webhooks that can dynamically change resources before admission
* Both have Validating Webhooks to configure what can/cannot go through admission
* Both provide a way to react to changes to pre-existing cluster resources (ie, resources that have already gone through admission)

Differences:

* Pepr is more like a "framework" than a tool. In Pepr you create a Pepr [Module](https://docs.pepr.dev/main/user*guide/pepr*modules/). In the Pepr module you define [capabilities](https://docs.pepr.dev/main/user*guide/capabilities/) that enforce / apply desired cluster state.
* Pepr is written in TypeScript. Kyverno is written in Go.
* Pepr provides the flexibility of a full-fledged, strongly typed programming language to decide what decisions to make based on events happening in the cluster. With Kyverno, you are limited to the constraints of YAML.
* Pepr can be used to reconcile events in order, similar to Kube-Builder or Operator SDK.

> * Pepr can apply a CustomResourceDefinition and control cluster state based on that custom resource. For example:
>   * When a `WebApplication` resource is created, deploy a certain deployment, service, serviceAccount, networkPolicy, serviceMonitor, PrometheusRule, etc.
>   * When a `WebApplication` resource is deleted, delete all the resources that were created for that `WebApplication`
>   * When a `WebApplication` is updated, update the deployed Kubernetes resources for that `WebApplication`

Author: ```suggestion
* Pepr can apply a CustomResourceDefinition and control cluster state based on that custom resource.
```

export NODE_OPTIONS="--disable-warning=DEP0040"
```

or

```bash
npx --node-options="--disable-warning=DEP0040" pepr [command]
```

## How does Pepr compare to Kyverno?

> Although Pepr and Kyverno have similarities, Pepr is very different than Kyverno.

Author: ```suggestion
Although Pepr and Kyverno have similarities, Pepr is very different than Kyverno. They have very different mission statements. Pepr focuses on making operators as easy as possible. Kyverno focuses on reporting, not building operators.
```

Similarities:

* Both have Mutating Webhooks that can dynamically change resources before admission
* Both have Validating Webhooks to configure what can/cannot go through admission

> * Both utilize Kubernetes Informers under the hood for pre-existing cluster resources (ie, resources that have already gone through admission)

Author: ```suggestion
* Both provide a way to react to changes to pre-existing cluster resources (ie, resources that have already gone through admission)
```

Differences:

> * They have very different mission statements. Pepr focuses on making operators as easy as possible. Kyverno focuses on reporting, not building operators.

Author: ```suggestion
```

npx --node-options="--disable-warning=DEP0040" pepr [command]
```

## How does Pepr compare to Kyverno?

Pepr has similarities but ultimately is very different than Kyverno.

> 

Author: mission statements are not the same.

Pepr is about making operators as easy as possible.

Kyverno is about reporting, not building operators

Reviewer: Added as first bullet point in differences section.

Differences:
* Pepr is more like a "framework" than a tool, in Pepr you create a Pepr [Module](https://docs.pepr.dev/main/user*guide/pepr*modules/), in the module itself you define [capabilities](https://docs.pepr.dev/main/user*guide/capabilities/) that enforce / apply desired cluster state
* Pepr is TypeScript and Kyverno is Go
* Pepr gives the flexibility of a full*fledged strongly typed programming language decide what decisions to make based on events happening in the cluster. IE, when this pod is created then i could theoretically create the same pod in another cluster. (You can do whatever you want because it is TypeScript)
* There is a side of Pepr that is much more similar to Kube*Builder or Operator SDK, it can be used to reconcile events in order. You can apply a CustomResourceDefinition and control cluster state based on that custom resource.
* * When I see a `WebApplication` resource is created I want to deploy a certain deployment, service, serviceAccount, networkPolicy, serviceMonitor, PrometheusRule, etc
* * When I see a certain `WebApplication` resource is deleted I wanted to delete said resources

> * * When I see it is updated ..

Author: ```suggestion
* * When I see it is updated, then update the deployed Kubernetes resources
```
